### PR TITLE
[Feature] Add CQL

### PR DESCRIPTION
**Pull request by BY571**
## Description

Implements the Conservative Q-Learning (CQL) objective for Offline Reinforcement Learning (https://arxiv.org/abs/2006.04779)
and adds examples for online and offline training. 

Performance tested for Pendulum:
image

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [x] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/objectives/cql.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+    """TorchRL implementation of the CQL loss.
+
+    Presented in "Conservative Q-Learning for Offline Reinforcement Learning" https://arxiv.org/abs/2006.04779
+
+    Args:
+        actor_network (ProbabilisticActor): stochastic actor
+        qvalue_network (TensorDictModule): Q(s, a) parametric model.
+            This module typically outputs a ``"state_action_value"`` entry.
+
+        loss_function (str, optional): loss function to be used with
```
Comment (vmoens) [inline torchrl/objectives/cql.py:53]:
```suggestion
    Keyword args:
        loss_function (str, optional): loss function to be used with
```

File: torchrl/objectives/cql.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+            the value function loss. Default is `"smooth_l1"`.
+        alpha_init (float, optional): initial entropy multiplier.
+            Default is 1.0.
```
Comment (vmoens) [inline torchrl/objectives/cql.py:56]:
```suggestion
            Default is ``1.0``.
```

File: torchrl/objectives/cql.py [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+
+    @property
+    def device(self) -> torch.device:
+        for p in self.parameters():
+            return p.device
+        raise RuntimeError(
+            "At least one of the networks of CQLLoss must have trainable " "parameters."
+        )
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
```
Comment (vmoens) [inline torchrl/objectives/cql.py:284]: See #1246 
It should not be too difficult to add `@dispatch` here. You will need to define the in_keys and out_keys.
Check SACLoss for instance, it's quite easy

File: torchrl/objectives/cql.py [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+            Default is ``False``.
+        deterministic_backup (bool, optional): Whether to use the deterministic. Default is ``True``.
+        num_random (int, optional): Number of random actions to sample for the CQL loss.
+            Default is ``10``.
+        with_lagrange (bool, optional): Whether to use the Lagrange multiplier.
+            Default is ``False``.
+        lagrange_thresh (float, optional): Lagrange threshold. Default is ``0.0``.
+        priority_key (str, optional): [Deprecated, use .set_keys(priority_key=priority_key) instead]
+            Tensordict key where to write the
+            priority (for prioritized replay buffer usage). Defaults to ``"td_error"``.
```
Comment (vmoens) [inline torchrl/objectives/cql.py:87]: Could you add a simple example?
Just copy paste the networks and loss construction from the tests, it should be sufficient.

**Review by vmoens — COMMENTED** [review]
Great work! Exciting stuff
A few comments

**Review by vmoens — COMMENTED** [review]
Great work! Exciting stuff
A few comments

**Comment by BY571** [comment]
I'm getting an error now regarding the in_keys of the objective, checking about that!

**Review by vmoens — APPROVED** [review]
